src/utils/cleeng.ts

```diff
@@ -1,11 +1,49 @@
 import type { AccessModel } from '../../types/Config';
+import type { MediaOffer } from '../../types/media';
+import type { PlaylistItem } from '../../types/playlist';
 
-export const isAllowedToWatch = (
-  accessModel: AccessModel,
-  isLoggedIn: boolean,
-  itemRequiresSubscription: boolean,
-  hasSubscription: boolean,
-): boolean =>
-  accessModel === 'AVOD' ||
-  (accessModel === 'AUTHVOD' && (isLoggedIn || !itemRequiresSubscription)) ||
-  (accessModel === 'SVOD' && (hasSubscription || !itemRequiresSubscription));
+/**
+ * The appearance of the lock icon, depending on the access model
+ *
+ * @param accessModel Platform AccessModel, excluding TVOD, which can only be applied per item
+ * @param isLoggedIn
+ * @param hasSubscription
+ * @param playlistItem Used to define if the item is 'free' or has mediaOffers
+ * @returns
+ */
+
+export const showLock = (accessModel: AccessModel, isLoggedIn: boolean, hasSubscription: boolean, playlistItem: PlaylistItem): boolean => {
+  const isItemFree = playlistItem?.requiresSubscription === 'false' || !!playlistItem?.free;
+  const mediaOffers = filterCleengMediaOffers(playlistItem?.productIds);
+
+  if (isItemFree) return false;
+  if (accessModel === 'AVOD' && !mediaOffers) return false;
+  if (accessModel === 'AUTHVOD' && isLoggedIn && !mediaOffers) return false;
+  if (accessModel === 'SVOD' && hasSubscription && !mediaOffers?.some((offer) => offer.premier)) return false;
+
+  return true;
+};
+
+/**
+ * Filters Cleeng MediaOffers from offers string
+ *
+ * @param offerIds String of comma separated key/value pairs, i.e. "cleeng:S916977979_NL, !cleeng:S91633379_NL, other_vendor:xyz123"
+ * Key is vendor, value is the offerId.
+ * Vendor keys starting with an exclamation mark represent a 'Premier Access' offer (TVOD only)
+ *
+ * @returns An array of MediaOffer { offerId, premier }
+ */
+export const filterCleengMediaOffers = (offerIds?: string): MediaOffer[] | null => {
```

Comment:
I've added `productIds` to `PlaylistItem`, since it's an optional property now. The MediaOffers (parsed) are not, so I don't think we should add them to PlaylistItem.
I've made it cleeng specific because it's the only key we 'care' about here. Not sure if we should make that more generic. Any more concrete ideas?

Reply:
What bothers me here is the repeated parsing for every lock and entitlement check. It's not a huge overhead, but it would be nice to do it once and remember it. The usages in the useEntitlement calls are all memoized but the showLock calls aren't.  We could do any of the following:

1. Make showLock into a hook and add memoization. Slightly more efficient, but still code to sync between showLock and useEntitlement
2. Combine showLock into useEntitlements with a param for whether or not to do the queries. Everything memoized and no code duplication, but params is kind of code smell.
3. Just call useEntitlements for everything and let memo and react query caching store results. Efficient and no duplication, but could be performance issues if lots of videos need querying.

Reply:
It's probably good to start decoupling the raw cleeng data formats from the data formats we use in the app logic anyhow.